### PR TITLE
Batched gemm

### PR DESCRIPTION
**Pull request by XapaJIaMnu**
When using MKL, we have access to `cblas_sgemm_batch`, which is _supposed_ to be more efficient than just having a for loop of `cblas_sgemm` calls.

### Checklist

- [x] I have tested the code manually
- [x] I have run regression tests
- [x] I have read and followed CONTRIBUTING.md
- [x] I have updated CHANGELOG.md

### CODE REVIEW

**Comment by emjotde** [comment]
Did you see any improvement? I tried this recently and it didn't do anything.

**Comment by XapaJIaMnu** [comment]
Here is a plot on a 'tiny' config preset (And using intgemm):
image

The batched sgemm is consistently slightly better.

**Comment by emjotde** [comment]
I am perplexingly unable to parse that graph even with the information in the title. :) What's x and y and why is one of them base-2 logarithmic?

**Comment by XapaJIaMnu** [comment]
Some disclaimer. The graph was given to me by intel. Y is time for the dataset to be translated, X is the number of cores. I'll try to get them to jump in here.

**Comment by sidkashyap-at-Intel** [comment]
> I am perplexingly unable to parse that graph even with the information in the title. :) What's x and y and why is one of them base-2 logarithmic?

This is not really a production graph :), let me explain: 
1) Y: time in seconds for Marian inference run
2) X: Number of threads on a Cascade Lake system 
3) Series: one with FP32 MKL cblas_sgemm_batch and another with FP32 MKL sgemm

**Comment by emjotde** [comment]
Wouldn't time in logarithmic scale always be quite close together unless there are huge improvements? If the goal was to show that batched gemm is faster this graph really doesn't do that :)

**Comment by sidkashyap-at-Intel** [comment]
> Wouldn't time in logarithmic scale always be quite close together unless there are huge improvements? If the goal was to show that batched gemm is faster this graph really doesn't do that :)

I did not defend that for a reason :)

**Comment by sidkashyap-at-Intel** [comment]
> > Wouldn't time in logarithmic scale always be quite close together unless there are huge improvements? If the goal was to show that batched gemm is faster this graph really doesn't do that :)
> 
> I did not defend the log2 axis for that reason :)
Better version of the graph here, [cblas_sgemm_batch](https://software.intel.com/en-us/mkl-developer-reference-c-cblas-gemm-batch) is slightly better than SGEMM. 

sgemm_vs_batch

File: src/tensors/cpu/prod.cpp [hunk 1]
```diff
@@ -134,6 +134,54 @@ void ProdBatched(marian::Tensor C,
...
+  CBLAS_TRANSPOSE transA_forarr = CblasNoTrans;
+  CBLAS_TRANSPOSE transB_forarr = CblasNoTrans;
+
+  if(transA)
+    transA_forarr = CblasTrans;
+
+  if(transB)
+    transB_forarr = CblasTrans;
+
+  static const constexpr size_t group_count = 1;
```
Comment (emjotde) [inline src/tensors/cpu/prod.cpp:147]: Please comment what this is and does.
Also please give a general comment how all these different vectors interact and what the batched gemm is supposed to do.

**Comment by emjotde** [comment]
Also please run the regression tests. The checklist is not really meant to be optional, especially for core functionality.

**Comment by emjotde** [comment]
And update CHANGELOG.md please

**Comment by XapaJIaMnu** [comment]
Hey,
@emjotde , I have added comments and documentation. I have run regression tests and they fail, however they fail in the same place in master too:

```
[2020-05-07 16:03:40] Total time: 2.85903s wall
+ python3 /home/dheart/uni_stuff/postdoc/marian-regression-tests/tools/sacrebleu/sacrebleu.py newstest2014.ref
+ tee model_base_fbgemm_packed8.out.bleu
BLEU+case.mixed+numrefs.1+smooth.exp+tok.13a+version.1.2.12 = 30.0 55.5/34.3/23.8/17.8 (BP = 1.000 ratio = 1.058 hyp_len = 2583 ref_len = 2442)
+ cat model_base_fbgemm_packed8.avx2.expected.bleu
BLEU+case.mixed+numrefs.1+smooth.exp+tok.13a+version.1.4.2 = 29.7 54.9/34.2/23.5/17.5 (BP = 1.000 ratio = 1.075 hyp_len = 2626 ref_len = 2442)
+ /home/dheart/uni_stuff/postdoc/marian-regression-tests/tools/diff.sh model_base_fbgemm_packed8.out model_base_fbgemm_packed8.avx2.expected
```

Cheers,

Nick

**Comment by emjotde** [comment]
We need to update the test, so not your problem.

**Pull request merged**